### PR TITLE
Demon Hunter creation implementation

### LINKED ISSUE #17642: [7.x | DB-Fix] added class_expansion_requirement for demon hunters
**Description**: adds expansion requirement datas for demon hunters to allow their creation and usage

**Current behaviour**: Demon hunter creation is currently disabled due to lack of creating database content.

**Expected behaviour**: Creation should work, I guess?

**Branch(es)**:  7.x 

```
DELETE FROM `class_expansion_requirement` WHERE `classID`= 12;
INSERT INTO `class_expansion_requirement` (`classID`, `expansion`) VALUES
(12, 6);
```

Todo: fix invalid class/race pair error --> need createinfo datas

rev.: 4587b5d88082d2c6416fafaa2f5b9f5f15038520

### PR DESCRIPTION
**Pull request by Ovahlord**
**Changes proposed**:
- added configurable requirements to create a demon hunter
- added database informations for required expansions
- added position datas for demon hunter starting zone (Zerixx for provided sniff datas)
- added rbac permission type to allow gamemasters to create demon hunters at any time

**Target branch(es)**: 6x

**Issues addressed**: Closes #17642

**Tests performed**: tested ingame, works as intended.

**Known issues and TODO list**:
- [x] need the sniffed value for orientation
- [x] need updated player levelstats for demon hunter BEFORE merging this pull request else the worldserver startup is gonna break due to missing database content (summoning @funjoker )

Images:
wowscrnshot_072216_143447
wowscrnshot_072216_143502
wowscrnshot_072216_143656

### CODE REVIEW

File: src/server/worldserver/worldserver.conf.dist [hunk 1]
```diff
@@ -812,6 +813,24 @@ HeroicCharactersPerRealm = 1
...
+#    DemonHuntersPerRealm
+#        Description: Limit number of demon hunter characters per account on this realm.
+#        Range:       1-10
+#        Default:     1
+
+DemonHuntersPerRealm = 1
+
+#
+#    CharacterCreating.MinLevelForDemonHunter
+#        Description: Limit creating heroic characters only for account with another
```
Comment (Aokromes) [inline src/server/worldserver/worldserver.conf.dist]: heroic > demon hunter :)

File: src/server/worldserver/worldserver.conf.dist [hunk 2]
```diff
@@ -854,6 +873,14 @@ StartPlayerLevel = 1
 StartHeroicPlayerLevel = 55
 
 #
+#    StartHeroicPlayerLevel
+#        Description: Staring level for heroic class characters after creation.
```
Comment (Aokromes) [inline src/server/worldserver/worldserver.conf.dist]: heroic  > demon hunter

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 3]
```diff
@@ -303,15 +304,23 @@ void WorldSession::HandleCharEnum(PreparedQueryResult result)
...
+
+            if (charInfo.Level >= sWorld->getIntConfig(CONFIG_CHARACTER_CREATING_MIN_LEVEL_FOR_DEMON_HUNTER))
                 charEnum.HasLevel70OnRealm = true;
 
             charEnum.Characters.emplace_back(charInfo);
         }
         while (result->NextRow());
     }
 
+    charEnum.IsDemonHunterCreationAllowed = (charEnum.HasDemonHunterOnRealm || !charEnum.HasLevel70OnRealm) ? false : true;
```
Comment (Shauren) [inline src/server/game/Handlers/CharacterHandler.cpp]: IsDemonHunterCreationAllowed is a separate criteria from HasDemonHunterOnRealm and HasLevel70OnRealm, on retail its used to allow early demon hunter access for people who preordered

File: src/server/game/World/World.cpp [hunk 4]
```diff
@@ -816,6 +816,15 @@ void World::LoadConfigSettings(bool reload)
 
     m_int_configs[CONFIG_CHARACTER_CREATING_MIN_LEVEL_FOR_HEROIC_CHARACTER] = sConfigMgr->GetIntDefault("CharacterCreating.MinLevelForHeroicCharacter", 55);
 
+    m_int_configs[CONFIG_DEMON_HUNTERS_PER_REALM] = sConfigMgr->GetIntDefault("DemonHuntersPerRealm", 1);
+    if (int32(m_int_configs[CONFIG_DEMON_HUNTERS_PER_REALM]) < 0 || m_int_configs[CONFIG_DEMON_HUNTERS_PER_REALM] > 10)
```
Comment (Shauren) [inline src/server/game/World/World.cpp]: there are 12 character slots per realm, not 10

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 5]
```diff
@@ -580,11 +589,11 @@ void WorldSession::HandleCharCreateCallback(PreparedQueryResult result, WorldPac
 
             _charCreateCallback.FreeResult();
 
-            if (!allowTwoSideAccounts || skipCinematics == 1 || createInfo->Class == CLASS_DEATH_KNIGHT)
+            if (!allowTwoSideAccounts || skipCinematics == 1 || createInfo->Class == CLASS_DEATH_KNIGHT || createInfo->Class == CLASS_DEMON_HUNTER)
             {
                 PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_SEL_CHAR_CREATE_INFO);
                 stmt->setUInt32(0, GetAccountId());
-                stmt->setUInt32(1, (skipCinematics == 1 || createInfo->Class == CLASS_DEATH_KNIGHT) ? 10 : 1);
+                stmt->setUInt32(1, (skipCinematics == 1 || createInfo->Class == CLASS_DEATH_KNIGHT || createInfo->Class == CLASS_DEMON_HUNTER) ? 10 : 1);
```
Comment (Shauren) [inline src/server/game/Handlers/CharacterHandler.cpp]: there are 12 character slots per realm, not 10

File: src/server/game/Handlers/CharacterHandler.cpp (hunk 3 above)
Comment (Ovahlord) [inline src/server/game/Handlers/CharacterHandler.cpp]: the client needs the iscreation allowed as well as false else the disbaled button will not show the reason why you cannot create the demon hunter (eg. no lvl 70 character). tested ingame befor did that change

File: src/server/worldserver/worldserver.conf.dist [hunk 6]
```diff
@@ -668,14 +668,15 @@ DeclinedNames = 0
...
+#        Default:     6 - (Expansion 6)
+#                     5 - (Expansion 5)
 #                     4 - (Expansion 4)
 #                     3 - (Expansion 3)
 #                     2 - (Expansion 2)
 #                     1 - (Expansion 1)
 #                     0 - (Disabled, Ignore and disable expansion content (maps, races, classes)
 
-Expansion = 5
+Expansion = 6
```
Comment (Shauren) [inline src/server/worldserver/worldserver.conf.dist]: not yet.

File: src/server/worldserver/worldserver.conf.dist [hunk 7]
```diff
@@ -812,6 +813,24 @@ HeroicCharactersPerRealm = 1
 CharacterCreating.MinLevelForHeroicCharacter = 55
 
 #
+#    DemonHuntersPerRealm
+#        Description: Limit number of demon hunter characters per account on this realm.
+#        Range:       1-10
```
Comment (Shauren) [inline src/server/worldserver/worldserver.conf.dist]: there are 12 character slots per realm, not 10

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 8]
```diff
@@ -303,15 +304,23 @@ void WorldSession::HandleCharEnum(PreparedQueryResult result)
...
+
+            if (charInfo.Level >= sWorld->getIntConfig(CONFIG_CHARACTER_CREATING_MIN_LEVEL_FOR_DEMON_HUNTER) || canAlwaysCreateDemonHunter)
                 charEnum.HasLevel70OnRealm = true;
 
             charEnum.Characters.emplace_back(charInfo);
         }
         while (result->NextRow());
     }
 
+    charEnum.IsDemonHunterCreationAllowed = (!charEnum.HasDemonHunterOnRealm || charEnum.HasLevel70OnRealm)
```
Comment (Shauren) [inline src/server/game/Handlers/CharacterHandler.cpp]: should have been &&

File: src/server/game/Entities/Player/Player.cpp [hunk 9]
```diff
@@ -11935,7 +11935,8 @@ void Player::VisualizeItem(uint8 slot, Item* pItem)
     if (pItem->GetTemplate()->GetBonding() == BIND_WHEN_EQUIPED || pItem->GetTemplate()->GetBonding() == BIND_WHEN_PICKED_UP || pItem->GetTemplate()->GetBonding() == BIND_QUEST_ITEM)
     {
         pItem->SetBinding(true);
-        GetSession()->GetCollectionMgr()->AddItemAppearance(pItem);
+        if (IsInWorld())
+            GetSession()->GetCollectionMgr()->AddItemAppearance(pItem);
```
Comment (Ovahlord) [inline src/server/game/Entities/Player/Player.cpp]: this change will not break the addition of the visualization. the appearances are still getting added without the messages after login.

**Comment by Shauren** [comment]
What about playercreateinfo action bars?

**Comment by Ovahlord** [comment]
need to know which spells are being in the bars and I can take care about it.

File: src/server/worldserver/worldserver.conf.dist [hunk 10]
```diff
@@ -854,6 +873,14 @@ StartPlayerLevel = 1
 StartHeroicPlayerLevel = 55
 
 #
+#    StartHeroicPlayerLevel
```
Comment (Naios) [inline src/server/worldserver/worldserver.conf.dist]: typo

File: src/server/worldserver/worldserver.conf.dist (hunk 1 above)
Comment (Bodeguero) [inline src/server/worldserver/worldserver.conf.dist]: @Ovahlord 
Nice work - I have some Pandaria scripts Draenor - if you want to look at them and try to implement I thank you.
PS: There is complete files some are just the base!

[Pandaria and Draenor](https://github.com/Ovahlord/TrinityCore/files/378445/Pandaria.and.Draenor.zip)

**Comment by Zerixx** [comment]
I can sniff this data for you in a bit

**Comment by Ovahlord** [comment]
@Zerixx would be awesome. try to catch the entire zone for spawns and so on. would speed up things by alot.

**Comment by Zerixx** [comment]
I will do my best.

**Comment by Zerixx** [comment]
I'll also let you know what's on the action bars

**Comment by Ovahlord** [comment]
@Zerixx and try to cast all spells of the class while your sniffer is attached~

**Comment by Shauren** [comment]
@Ovahlord but please do not add that into this PR - just waiting for action bars to be sniffed and added here and thats it, no more things here.

File: src/server/scripts/Spells/spell_dh.cpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+#include "Containers.h"
+
+enum DemonHunterSpells
+{
+    SPELL_CHAOS_STRIKE_MAINHAND = 199547,
+    SPELL_CHAOS_STRIKE_OFFHAND  = 222031,
+};
+
+// 197125 - Chaos Strike
+class spell_dh_chaos_strike : public SpellScriptLoader
```
Comment (Shauren) [inline src/server/scripts/Spells/spell_dh.cpp]: move this to separate PR

File: src/server/scripts/Spells/spell_dh.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    class spell_dh_chaos_strike_AuraScript : public AuraScript
+    {
+        PrepareAuraScript(spell_dh_chaos_strike_AuraScript);
+
+        void HandleEffectProc(AuraEffect const* aurEff, ProcEventInfo& eventInfo)
+        {
+            PreventDefaultAction();
+
+            if (Unit* caster = GetCaster())
+                if (eventInfo.GetSpellInfo() && eventInfo.GetHitMask() & PROC_HIT_CRITICAL && (eventInfo.GetSpellInfo()->Id == SPELL_CHAOS_STRIKE_MAINHAND ||
```
Comment (Shauren) [inline src/server/scripts/Spells/spell_dh.cpp]: you never worked with proc system, did you?

Comment (Ovahlord) [inline src/server/scripts/Spells/spell_dh.cpp]: not that often, ye

Comment (Shauren) [inline src/server/scripts/Spells/spell_dh.cpp]: you control these condtions - spell, hit type with fields in database

**Comment by PiskelDEV** [comment]
Love you guys, good job :+1:

**Comment by Zerixx** [comment]
Sniffing the entirety of the DH starting.. even vault.

File: src/server/worldserver/worldserver.conf.dist [hunk 13]
```diff
@@ -668,7 +668,8 @@ DeclinedNames = 0
 #    Expansion
 #        Description: Allow server to use content from expansions. Checks for expansion-related
 #                     map files, client compatibility and class/race character creation.
-#        Default:     5 - (Expansion 5)
+#        Default:     6 - (Expansion 6)
```
Comment (Warpten) [inline src/server/worldserver/worldserver.conf.dist]: you forgot to actually change that a few lines below.

File: src/server/worldserver/worldserver.conf.dist [hunk 14]
```diff
@@ -812,6 +813,24 @@ HeroicCharactersPerRealm = 1
...
+
+#
+#    CharacterCreating.MinLevelForDemonHunter
+#        Description: Limit creating demon hunters only for account with another
+#                     character of specific level.
+#        Default:     70 - (Enabled, Requires at least another level 70 character)
+#                     0  - (Disabled)
+#                     1  - (Enabled, Requires at least another level 1 character)
+
+CharacterCreating.MinLevelForDemonHunter = 70
```
Comment (Warpten) [inline src/server/worldserver/worldserver.conf.dist]: Even though DH are an "heroic" class, there's no such requirement as far as we knew (and beta confirmed that, unless that was only temporary - then again, blizzard hasnt statuated on that point)

File: src/server/worldserver/worldserver.conf.dist (hunk 13 above)
Comment (Ovahlord) [inline src/server/worldserver/worldserver.conf.dist]: see comment of shauren "not yet".

File: src/server/worldserver/worldserver.conf.dist (hunk 14 above)
Comment (Ovahlord) [inline src/server/worldserver/worldserver.conf.dist]: It is in the client and so I think that this is going to be true that there is a level requirement. If it's a false implementation it can be dropped easily. According to wowwiki and forums the level 70 cap is intended http://wowwiki.wikia.com/wiki/Demon_Hunter

**Comment by Zerixx** [comment]
Here ya go. Almost the entire DH starting bit.. it DC'd be at the last minute. But it should have all you need.. Sorry I have a few packets I have not updated yet for WPP.

https://www.dropbox.com/s/t3jkyt4we95qdog/DH%20Sniffs.rar?dl=0

File: src/server/scripts/Spells/spell_dh.cpp [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+ * Ordered alphabetically using scriptname.
+ * Scriptnames of files in this file should be prefixed with "spell_dk_".
+ */
+
+#include "Player.h"
+#include "ScriptMgr.h"
+#include "SpellScript.h"
+#include "SpellAuraEffects.h"
+#include "SpellHistory.h"
+#include "Containers.h"
```
Comment (Vincent-Michael) [inline src/server/scripts/Spells/spell_dh.cpp]: current no need for headers

**Comment by Ovahlord** [comment]
the coordinates have been corrected. the action bars have been correct and so the database part for the creation is complete. Now there is just the levelstats left

**Comment by Zerixx** [comment]
@Ovahlord This is probably close. https://gist.github.com/funjoker/d4d4133538a49d6394a2940e825f3e22

**Comment by Ovahlord** [comment]
@Zerixx it is incomplete. Level 98 and 99 are missing and the core also wants level 1 - 97 which means the db needs stats from 1 - 97 with 0 in all fields.

**Comment by Ovahlord** [comment]
sure but I think those stats include the stats of your equipment which might cause inaccurate stats.

**Comment by Zerixx** [comment]
That's a good point. If you need anything else from me, just hmu.

**Comment by funjoker** [comment]
They are in now

**Comment by Ovahlord** [comment]
@Zerixx did you do any double jump while doing that sniff? I cannot find any spellcast related to it in your sniff. Would need one with at least one double jump in it

**Comment by Zerixx** [comment]
I used double jump a ton. @Ovahlord

**Comment by Ovahlord** [comment]
@Zerixx  very interesting. Would you mind doing a very very small sniff by logging in, doing a double jump and then closing the client directy? I think that would help me quite alot to find out whats going on

**Comment by Zerixx** [comment]
Believe it or not, my sniffer legitimately just f'd up and wont find ProcessMessage. Once I figure out why, I'll do the sniff for you. 

@Shauren, any idea why ^ may have happened?

**Comment by Shauren** [comment]
@Zerixx no, I don't rely on that for wstc.

**Comment by Zerixx** [comment]
@Shauren, no worries. Was just curious if you knew anything about it. But, yeah, @Ovahlord once I fix up the issue in the sniffer, I'll get that for you.

**Comment by Ovahlord** [comment]
@Zerixx u dont have to do it anymore. I found the handling already. Might take a day or two but it will be working

File: src/server/worldserver/worldserver.conf.dist [hunk 16]
```diff
@@ -796,7 +797,7 @@ CharactersPerRealm = 12
 #
 #    HeroicCharactersPerRealm
 #        Description: Limit number of heroic class characters per account on this realm.
-#        Range:       1-10
+#        Range:       1-12
 #        Default:     1
```
Comment (Exodius) [inline src/server/worldserver/worldserver.conf.dist]: Shouldn't you change `Default:     1` to `Default:     1`? DKs and DHs are both Heroic Characters. If `Default=1`, shouldn't that prevent players from creating DH if DK is already created and vice-versa?

Comment (tkrokli) [inline src/server/worldserver/worldserver.conf.dist]: If you check in Shauren's commit, https://github.com/TrinityCore/TrinityCore/commit/79aafd5865f11b5ecdd2865829562b981652db75#diff-cb0c3fe3d6f6147fc1a95217aff13561L800 , you will see that he did not change it either. Maybe you could propose that change in a new PR only for worldserver.conf.dist (unless it is added in one of the upcoming PRs before Legion is live from 2016-08-30).

Comment (Exodius) [inline src/server/worldserver/worldserver.conf.dist]: I **could** but, if I am not mistaken, you would need to separate these checks...
Just changing `Default: 1` to `Default: 2` would enable people to create 2 DKs or 2 DHs. This config (`HeroicCharactersPerRealm`) needs to be redesigned and split into two separate ones. Example:

```
DeathKnightCharactersPerRealm
Description: Limit number of Death Knight characters per account on this realm.
Range:       1-12
Default:     1
```

```
DemonHunterCharactersPerRealm
Description: Limit number of Demon Hunter characters per account on this realm.
Range:       1-12
Default:     1
```

Comment (Exodius) [inline src/server/worldserver/worldserver.conf.dist]: I don't know WTF am I writing about... I just re-checked the commit and saw that there IS already `DemonHuntersPerRealm` present.

Comment (tkrokli) [inline src/server/worldserver/worldserver.conf.dist]: You got a point there. It needs some extra coding too, I guess.

Comment (Ovahlord) [inline src/server/worldserver/worldserver.conf.dist]: It's a simple renaming. Feel free to do a pullrequest~

**Comment by DrEhsan** [comment]
@Zerixx if you havent fixed your issue in sniffer , you can use this :https://github.com/DrEhsan/Whiff
i updated to the x86 builds.

thanks to @Zedron for his sniffer

**Comment by Zerixx** [comment]
@DrEhsan Thanks, man.. got my mine working though.

**Comment by searum** [comment]
Zerixx, is there the sql file from sniff too?
